contracts/pcv/compound/CompoundBadDebtSentinel.sol

```diff
@@ -0,0 +1,259 @@
+
+        for (uint256 i = 0; i < accountsLength; i++) {
+            (, , uint256 badDebt) = IComptroller(comptroller)
+                .getAccountLiquidity(addresses[i]);
+            totalBadDebt += badDebt;
+        }
+    }
+
+    /// ------------- Public State Changing API -------------
+
+    /// @notice rescue funds from all stored PCV Deposits
+    /// no need for the reentrancy lock as the PCV Guardian will lock when withdrawals process
+    /// @param addresses of compound users to query for bad debt
+    function rescueAllFromCompound(
+        address[] memory addresses
+    ) external override {
+        _countBadDebtAndWithdraw(addresses, allPcvDeposits());
+    }
+
+    /// @notice rescue assets from compound, then pause all deposits
+    /// no need for the reentrancy lock as the PCV Guardian will lock when withdrawals process
+    /// @param addresses of compound users to query for bad debt
+    /// @param pcvDeposits to pull funds from
+    function rescueFromCompound(
+        address[] memory addresses,
+        address[] memory pcvDeposits
+    ) external override {
+        uint256 pcvDepositsLength = pcvDeposits.length;
+
+        /// validate pcv deposits are ordered and not duplicated
+        require(
+            noDuplicatesAndOrdered(pcvDeposits),
+            "CompoundBadDebtSentinel: PCVDeposits, Duplicates or OOO"
+        );
+
+        /// validate pcv deposits are in whitelist
+        unchecked {
+            for (uint256 i = 0; i < pcvDepositsLength; i++) {
+                require(
+                    isCompoundPcvDeposit(pcvDeposits[i]),
+                    "CompoundBadDebtSentinel: Invalid Compound PCV Deposit"
+                );
+            }
+        }
+
+        _countBadDebtAndWithdraw(addresses, pcvDeposits);
+    }
+
+    /// @notice helper method that validates no duplicate user addresses are passed in
+    /// @param addresses of compound users to query for bad debt
+    /// @param pcvDeposits to pull funds from
+    function _countBadDebtAndWithdraw(
+        address[] memory addresses,
+        address[] memory pcvDeposits
+    ) private {
+        require(
+            noDuplicatesAndOrdered(addresses),
+            "CompoundBadDebtSentinel: Addresses, Duplicates or OOO"
+        );
+
+        uint256 pcvDepositsLength = pcvDeposits.length;
+
+        /// figure out how much bad debt exists
+        /// at this point, we know there are no duplicate accounts in the address array
+        uint256 totalBadDebt = getTotalBadDebt(addresses);
+
+        if (totalBadDebt >= badDebtThreshold) {
+            unchecked {
+                for (uint256 i = 0; i < pcvDepositsLength; i++) {
+                    /// morpho reverts on withdrawing 0 balance
+                    /// so do not withdraw on 0 balance
+                    if (PCVDeposit(pcvDeposits[i]).balance() != 0) {
+                        try
+                            PCVGuardian(pcvGuardian).withdrawAllToSafeAddress(
```

Comment:
update pcv guardian to allow withdrawing all available liquidity or balance, whichever is less